src/dataflow/src/render/sources.rs

```diff
@@ -63,6 +63,67 @@ impl<G> Context<G, Row, Timestamp>
 where
     G: Scope<Timestamp = Timestamp>,
 {
+    /// Render a local source and return a handle to its input, and the data and
+    /// error collections.
+    fn render_local_source(
+        &mut self,
+        render_state: &mut RenderState,
+        scope: &mut G,
+        persisted_name: Option<String>,
+    ) -> (LocalInput, Collection<G, Row>, Collection<G, DataflowError>) {
+        let ((handle, capability), ok_stream, err_collection) = {
```

Comment:
nit: for bonus points, you could actually split up this part, that deals with handles, and the persist part. It's not obvious in the code, but a local source with a `persisted_name` is completely different from a local source without one.

Could be `render_local_handle_source()` and `render_local_persistence_source()`, for example.